File: Problems/RansomNote/solutions.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
  //   Ransom Note: Iterate over letters and create a hash map of the letters and occurence: <key: char> : <count: int>
  let ransom_map: { [key: string]: number } = {};
  for (let ransom_letter of ransomNote) {
    // Check if letter(key) is not in the hashmap
    if (!ransom_map[ransom_letter]) {
      ransom_map[ransom_letter] = 1;
    } else {
      // if it is... increase the count by 1
      ransom_map[ransom_letter] += 1;
    }
  }

  //   Magazine: Iterate over letter and create a hash map of the letters and occurence <key: char> : <count: int>
  let magazine_map: { [key: string]: number } = {};
  for (let mag_letter of ransomNote) {
    // Check if letter(key) is not in the hashmap
    if (!ransom_map[mag_letter]) {
      ransom_map[mag_letter] = 1;
    } else {
      // if it is... increase the count by 1
      magazine_map[mag_letter] += 1;
    }
  }

  //   Check ransom map keys & count against the magazine map keys & count
  for (const key in ransom_map) {
    if (
      magazine_map[key] === undefined ||
      magazine_map[key] < ransom_map[key]
    ) {
      return false;
    }
  }
  return true;
}

const ransomNote = "aa";
const magazine = "aab";

let ans = canConstruct(ransomNote, magazine);

File: Problems/General/RansomNote/solutions.py
# def canConstruct(ransomNote: str, magazine: str) -> bool:
#     from collections import Counter
#     # ransomNote_letter_map = {l: ransomNote.count(l) for l in ransomNote}
#     ransomNote_letter_map = Counter(ransomNote)

#     # magazine_letter_map = {l: magazine.count(l) for l in magazine}
#     magazine_letter_map = Counter(magazine)
#     for key in ransomNote_letter_map:
#         if key not in magazine_letter_map or magazine_letter_map[key] < ransomNote_letter_map[key]:
#             return False
#     return True

def canConstruct(ransomNote: str, magazine: str) -> bool:
    ransom_map = {}
    magazine_map = {}

    for rNote_letter in ransomNote:
        if rNote_letter not in ransom_map:
            ransom_map[rNote_letter] = 1
        else:
            ransom_map[rNote_letter] += 1

    for mag_letter in magazine:
        if mag_letter not in magazine_map:
            magazine_map[mag_letter] = 1
        else:
            magazine_map[mag_letter] += 1

    for key in ransom_map:
        if key not in magazine_map or magazine_map[key] < ransom_map[key]:
            return False
    return True


if __name__ == "__main__":
    ransomNote = "itwqbtcdprfsuprkrjkausiterybzncbmdvkgljxuekizvaivszowqtmrttiihervpncztuoljftlxybpgwnjb"
    magazine = "noclrnthqqganujpsabdjjrwfljribdbwoskcnmtxokkrlvcvsdcxlcqtmugwdedfebnveivovwmkcgncoqslqteecmnekxqkaguebafmmldaljsdhfjasdhljshfuweryhjfasdjfg"
    out = canConstruct(ransomNote=ransomNote, magazine=magazine)
    print(out)
